plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployJob.java

```diff
@@ -168,54 +169,175 @@ protected void canceling() {
+              | IWorkbenchBrowserSupport.NAVIGATION_BAR | IWorkbenchBrowserSupport.STATUS;
+          browserSupport.createBrowser(style, project, project, project).openURL(url);
+        } catch (PartInitException ex) {
+          // Unable to use the normal browser support, so push to the OS
+          logger.log(Level.WARNING, "Cannot launch a browser", ex);
+          Program.launch(finalAppLocation);
+        } catch (MalformedURLException ex) {
+          return StatusUtil.error(getClass(), Messages.getString("invalid.url"), ex);
+        }
+        return Status.OK_STATUS;
+      }
+
+    };
+    launchBrowserJob.schedule();
+    return Status.OK_STATUS;
+  }
+
+  /**
+   * @return the error message obtained from <code>config.getErrorMessageProvider()</code> or
+   * <code>defaultMessage</code>
+   */
+  private String getErrorMessageOrDefault(String defaultMessage) {
+    // TODO: Check the assumption that if there are error messages during staging collected via
+    // the errorCollectingLineListener, the staging process will have a non-zero exitcode,
+    // making it ok to use the same errorCollectingLineListener for the deploy process
+    List<String> messages = errorCollectingLineListener.getCollectedMessages();
+    if (!messages.isEmpty()) {
+      return Joiner.on('\n').join(messages);
+    } else {
+      return defaultMessage;
+    }
+  }
+
+  private String getDeployedAppVersion() {
+    String rawDeployOutput = stdoutLineListener.getOutput().substring(stagingOutputEndIndex);
+    DeployOutput deployOutput = AppEngineDeployUtil.parseDeployOutput(rawDeployOutput);
+    return deployOutput.getVersion();
+  }
+
   private final class StoreProcessObjectListener implements ProcessStartListener {
     @Override
     public void onStart(Process proces) {
       process = proces;
     }
   }
 
-  private final class RecordProcessError implements ProcessExitListener {
-    // temporary way of error handling, after #439 is fixed, it'll be cleaner
+  private class StagingExitListener implements ProcessExitListener {
+    private IStatus status;
+
+    @Override
+    public void onExit(int exitCode) {
+      if (cloudSdkProcessStatus == Status.CANCEL_STATUS) {
+        status = cloudSdkProcessStatus;
+      } else if (exitCode != 0) {
+        // temporary way of error handling, after #439 is fixed, it'll be cleaner
+        String errorMessage =
+            getErrorMessageOrDefault(Messages.getString("cloudsdk.process.failed", exitCode));
+        status = StatusUtil.error(this, errorMessage);
+      } else {
+        stagingOutputEndIndex = stdoutLineListener.getOutput().length();
+        status = Status.OK_STATUS;
+      }
+    }
+
+    /**
+     * @return the status on exit of the process or null is the process has not exited.
+     */
+    public IStatus getExitStatus() {
+      return status;
+    }
+  }
+
+  private final class DeployExitListener implements ProcessExitListener {
```

Comment:
There's a lot of duplicate code in all these exit listeners. Any ideas how we can reduce that?